### DISCUSSION TITLE
How to define DiracKernel that is a function in space and time

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#20838] [Q&A General] [answered]
Hi

I am trying to simulate an ultrasound push inside soft tissue. The push spatial distribution and its variation in time are given. I used to use `ReporterPointSource` to define point sources distributed in space. Now, how can I define the variation in time of those sources?

**Comment by GiudGiud** [comment, marked as answer]
Hello

To make them vary in time you ll have to vary the reporter values. 
What kind of reporter are you using?
I think you essentially need a vectoroffunctionpostprocessor but that doesnt exist rn. You can make one with this:
https://mooseframework.inl.gov/source/postprocessors/FunctionValuePostprocessor.html
https://mooseframework.inl.gov/source/vectorpostprocessors/VectorOfPostprocessors.html

Guillaume
Guillaume

**Reply by aaelmeli** [reply]
> 
Thanks, @GiudGiud 
I am using the CSVReader. I will look into those options.

**Reply by GiudGiud** [reply]
CSVReader doesnt do time dependence sadly.

**Reply by aaelmeli** [reply]
> CSVReader doesnt do time dependence sadly.

So, generally, can we model a force that is varying in time and has user-defined spatial distribution through moose? 
The force that I have is an impulse or a step function in time and has a spatial distribution that is given (at each node). Can we model this in moose?

**Reply by GiudGiud** [reply]
Do you have a parsed expression for the function? If so that's the easiest. 
How many time steps is there? If a small number we can hack something together with parsed expressions (if(t<t1, use function_1, else if t<t2, etc) easily. 

Do you have the data only in a CSV format?
Other options could be pre-loaded in an Exodus file.
In the CSV file can the data be easily separated in different files for each time step?

**Reply by aaelmeli** [reply]
For the ideal case, a parsed expression for the spatial distribution is not available. But I can figure out a simplified version of the force so that I can have parsed expression for it.
The data is available on a grid produced by a matlab script. So, I think it can be written in different formats.

Yes, I can separate the data for each time step in the CSV file.

**Reply by GiudGiud** [reply]
Ok I need to sleuth for a bit through the docs to figure out what s the best option from CSV. 
I think easiest will be to read a CSV file on every time step

**Reply by GiudGiud** [reply]
Actually, these two functions can do time and space dependent CSV files
https://mooseframework.inl.gov/source/functions/PiecewiseMultilinear.html
https://mooseframework.inl.gov/source/functions/PiecewiseMulticonstant.html

from there you d need to link them up to the Dirac Kernels. 
I think this one should work, but you ll need multiple of them
https://mooseframework.inl.gov/source/dirackernels/FunctionDiracSource.html

To work with the ReporterPointSource,
You could:
- modify FunctionDiracSource to have multiple points
OR
- modify FunctionValuePostprocessor to chose the point to evaluate the function at
OR
- write a FunctionValuesVectpostprocessor to evaluate a given function at multiple points, specified in either a CSV file or with x/y/z vectors

Either of these options would be good to add to MOOSE

**Reply by aaelmeli** [reply]
Thanks @GiudGiud 
Actually, I found this `UserForcingFunctionNodalKernel`
https://mooseframework.inl.gov/source/nodalkernels/UserForcingFunctionNodalKernel.html
which I think will work fine combined with `PiecewiseMulticonstant` functions. The only issue would be that the user has to define the force values on all grid points. For example, if the user wanted to model only one point source in one location (let's say in the middle of the domain), the user must set zero force to all other points in the grid.

**Reply by GiudGiud** [reply]
This feels like a very different object, so I m not sure. But if you say so then I m glad you have what you need!
